### ISSUE TITLE
Update core vic-machine code to allow authentication using a session key, session cookie, or SAML token

### ISSUE THREAD

**Issue by zjs** [vmware/vic#6033]
Update core `vic-machine` code to allow authentication using a SAML token (preferably) or some other session credential available to the vSphere H5 client (such as a session key or session cookie) in place of a username and password.

Eventually, this will be exposed by the API as well (#6034).

**Comment by zjs** [comment]
To capture some additional information:

 * Based on information provided by @jooskim, it looks like we have access to a SAML Holder-of-Key Token, but _not_ the corresponding private key. This may rule out SAML authentication.
 * We do have access to a username, session key, and session cookie. Based on information provided by @dougm, it sounds like this is similar to what we get back when we log in using the username and password. Therefore, we should be able to use these in place of a username and password when interacting with vSphere on the user's behalf.

**However**, we don't have anything that we can persist when creating the VCH so that _it_ can interact with vSphere (the session key/cookie will expire, at which point the VCH would stop working, which would be... bad). One option is to require that the user always supply operations user credentials when creating a VCH via the API (even if those credentials are for the user's own account). This is a decent _short term_ option, so that the rest of this Epic is not blocked. Much better would be to create a minimally-privileged service account for the VCH as a part of its creation (which is tracked by #635, which @hickeng notes may be very out-of-date at this point).

I'd propose re-purposing this issue to track gathering additional information, proposing updates to the design document, and then reflecting those changes in GitHub (in this Epic and/or #635, as appropriate).

**Comment by jonathanibers** [comment]
So investigating here and #635, I had some questions about this story.

According to #2875, vic-machine currently has support for a seperate user, specified by the --ops-user and --ops-password options.  Looking at the code, it's used in create, and if not specified, vic-machine uses the user/pass from --username and --password.

There is documentation about using --ops-user [https://github.com/vmware/vic/blob/08-docs/doc/user_doc/vic_installation/vch_installer_examples.md#ops-user](link), and vic-machine-linux already accepts an ops-user.   By the documentation, there is a process to create this ops user with fewer priviledges.

But it looks like the ops-user account has to be set before creation, and it seems like the only way to create it is through the UI in vsphere.  Do we want this story to be about creating this ops account in vic-machine, and / or have this created automatically when creating a VCH?  So when the admin calls vic-machine create, it prompts for a user/password to be used for that VCH that becomes our ops user?

Thanks for any help!

@zjs @hickeng

**Comment by zjs** [comment]
>  Do we want this story to be about creating this ops account in vic-machine, and / or have this created automatically when creating a VCH? So when the admin calls vic-machine create, it prompts for a user/password to be used for that VCH that becomes our ops user?

I think that automatically creating a user account when creating the VCH should be our goal. However, I think we should create these as service accounts (also call "solution users" or "application users").

When creating a solution user in vSphere, I believe we supply a certificate instead of a password. We would store the corresponding private key with the VCH (in a similar way to how we store the password today). This is a bit more secure, and avoids the need to have the administrator supply anything.

We can continue to support a user-supplied `--ops-user` that overrides this behavior for customers who want to manage the user account themselves.

I'm on PTO, and @hickeng will be traveling. Maybe @lcastellano could help with any questions in our absence?

**Comment by jonathanibers** [comment]
Thanks for the reply!

So are these service accounts separate from the ops accounts?  What differentiates them?

And is there documentation somewhere about how we store accounts in the VCH (and / or where in the codebase would that be)?

**Comment by jonathanibers** [comment]
So #635 is tracking creating these minimally-priviledged service users, and I've been diving through both that and #716 to get a handle on this.  Based on what I see, the permissions for different user roles was discussed in #1689, which is now in the documentation [here](https://vmware.github.io/vic-product/assets/files/html/1.1/vic_vsphere_admin/set_up_ops_user.html).  These different roles are for ops users that are created (currently) within the UI.

@andrewtchin had a WIP to programmatically create these limited permissions users [here](https://github.com/andrewtchin/vic/tree/4444/ops-user), and for the token-based authentication @dougm had a WIP to issue a token [here](https://github.com/vmware/govmomi/compare/master...dougm:sts) and tracked in #766, but mentioned the govmomi LoginByToken method was more complicated than initially thought.  (Though this was 4 months ago so I don't know if this has changed since then).

Is there an example for how we would want a user and/or admin to use these service accounts?  Currently (as I understand it) we have an admin deploy the VCH, and specifies an --ops-user if they desire to have a separate, limited operations user (that they have created beforehand).  This ops-user can perform only tasks defined by whatever roles it was created with.

In an ideal world, would it work something like:  admin creates a VCH.  As part of it's creation, the VCH creates a service account, stores private key, and returns a certificate, that can be used to operate said VCH with limited permissions.  The limited operations user can now use vic-machine to configure and manage the VCH, supplying this certificate instead of a username/password.  This certificate is checked against the private key, which defines the operations that the user has access to.

Trying to wrap my head around 1) what functionality we have now and 2) what functionality we want.  Thanks for any help / clarifications!

@zjs (for when you get back from PTO :) )

**Comment by zjs** [comment]
After reading this a few times, I think I understand the source of confusion during the standup. 

The operations user account isn't used when calling `vic-machine` and won't be used for managing the VCH through the UI. It is used _by the VCH_, when the VCH needs to interact with vSphere.

To summarize the behavior today, we have two cases:

1. A VCH is created without specifying an operations user.
   -  When creating the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to create the VCH. These credentials are also stored (safely and securely) with the VCH.
   - When inspecting the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to retrieve the necessary information. 
   - When reconfiguring the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to make the necessary changes. 
   - When creating a container within the VCH, a container app developer authenticates to the VCH using their client certificate. In this case, the developer is interacting with the VCH via the docker client and no vSphere credentials are supplied with the operation. The VCH uses the vSphere Administrator's stored credentials to interact with vSphere to deploy the container.
2. A VCH is created while specifying an operations user.
   -  When creating the VCH, a vSphere Administrator supplies their username and password to `vic-machine`, as well as the username and password for a separate "operations" account. The administrator's credentials are used to interact with vSphere to create the VCH. The operations credentials are also stored (safely and securely) with the VCH.
   - As in case 1: When inspecting the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to retrieve the necessary information. 
   - As in case 1: When reconfiguring the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to make the necessary changes. 
   - When creating a container within the VCH, a container app developer authenticates to the VCH using their client certificate. In this case, the developer is interacting with the VCH via the docker client and no vSphere credentials are supplied with the operation. The VCH uses the stored operations credentials to interact with vSphere to deploy the container.

(@stuclem: Maybe we can find a way to incorporate some of the above into [the relevant page of the docs](https://vmware.github.io/vic-product/assets/files/html/1.1/vic_vsphere_admin/set_up_ops_user.html)?)

Storing the vSphere Administrator's credentials is undesirable, and problematic. (For example: what happens when they change their password?)

In the case of the UI (and API), storing the vSphere Administrator's credentials isn't feasible: the UI plugin doesn't have access to the user's password (which is a good thing, from a security point of view).

So the desired model is:  

1. A VCH is created without specifying an operations user.
   -  When creating the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to create the VCH. These credentials are used to create a service account with a minimal set of permissions. The _service account_'s credentials (i.e., private key) are stored (safely and securely) with the VCH.
   - As today, when inspecting the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to retrieve the necessary information. 
   - As today, when reconfiguring the VCH, a vSphere Administrator supplies their username and password to `vic-machine`. These credentials are used to interact with vSphere to make the necessary changes.
   - When viewing or managing the VCH via the UI, the vSphere Administrator's session is passed to the API, and that session is used to interact with vSPhere to retrieve the necessary information or make the necessary changes.
   - When creating a container within the VCH, a container app developer authenticates to the VCH using their client certificate. In this case, the developer is interacting with the VCH via the docker client and no vSphere credentials are supplied with the operation. The VCH uses the stored _service account_'s credentials to interact with vSphere to deploy the container.
2. A VCH is created while specifying an operations user.
   - Behaves like today.

To summarize, the operations credentials are used by the VCH when it interacts with vShere (e.g,. as a result of a docker command), not when interacting `vic-machine` or the new UI. The key change in functionality is to automatically create an operations user for each VCH by default, and to adhere to the best practices for service/automation accounts when doing so.

**Comment by jonathanibers** [comment]
That was super helpful and cleared a lot of things up.  Thanks Zach!

So what would be the next steps here?  I'm currently following the instructions in the documentation for ops users and looking at the codepath to figure out how it works.  Is there something else I should be working on as part of this story?

**Comment by jonathanibers** [comment]
WIP on this issue linked below.  I'm moving to PKS, but I had some work done in trying to implement SSO for the VCH.

https://github.com/jonathanibers/vic/tree/sso-wip
https://github.com/jonathanibers/govmomi/tree/sso-wip

**Comment by jzt** [comment]
Progress report:

[This proof-of-concept change](https://github.com/jzt/vic/commit/3d46e9d5145c8e959192d0766300d0707bf8a8b5) allows us to set the session key cookie by passing it in a header. It works if I make an initial request using basic HTTP auth, printing the session key and then passing it into a subsequent request in a header. However, it doesn't work with a session key taken from a browser login (as I initially suspected). This poses a problem for us in that we only have the browser session cookie when managing a VCH through the H5 plugin.

basic auth:
```
$ curl -s -w "%{http_code}\n" -X GET "http://127.0.0.1:31337/container/target/192.168.1.8:443/datacenter/ha-datacenter/vch?thumbprint=40:B0:7C:C0:C7:4A:B4:6D:39:DC:DD:97:C5:85:DD:53:4C:06:1D:02" -H  "authorization: Basic cm9vdDpwYXNzd29yZA=="
{"vchs":[{"admin_portal":"https://192.168.1.11:2378","docker_host":"192.168.1.11:2376","id":"96","name":"vic-docker","upgrade_status":"Up to date","version":"v1.2.0-rc1-0-db19ca4"}]}
200
```
server prints the key:
```
$ bin/vic-machine-server --port 31337 --scheme http
2017/10/13 14:21:34 Serving vic machine at http://127.0.0.1:31337
INFO[0016] Key: c5953a0513d6ae2577025ff3c62ebc14a183f993
INFO[0002] Validating target
```
pass key into header in subsequent request: 
```
$ curl -s -w "%{http_code}\n" -X GET "http://127.0.0.1:31337/container/target/192.168.1.8:443/vch?thumbprint=40:B0:7C:C0:C7:4A:B4:6D:39:DC:DD:97:C5:85:DD:53:4C:06:1D:02" -H  "X-VMWARE-SESSION: c5953a0513d6ae2577025ff3c62ebc14a183f993"
{"vchs":[{"admin_portal":"https://192.168.1.11:2378","docker_host":"192.168.1.11:2376","id":"96","name":"vic-docker","upgrade_status":"Up to date","version":"v1.2.0-rc1-0-db19ca4"}]}
200
```
attempt with browser key:
```
$ curl -s -w "%{http_code}\n" -X GET "http://127.0.0.1:31337/container/target/192.168.1.8:443/vch?thumbprint=40:B0:7C:C0:C7:4A:B4:6D:39:DC:DD:97:C5:85:DD:53:4C:06:1D:02" -H  "X-VMWARE-SESSION: 79103e002950f4ec2c5b69df570d571941199a3f"
{"message":"Target validation failed: validation of configuration failed"}
400
```
server no likey:
```
ERRO[0039] --------------------
ERRO[0039] Failed to get host system: please specify a datacenter
```

@zjs @dougm Can you think of anything I might be doing incorrectly here?

**Comment by zjs** [comment]
> ```ERRO[0039] Failed to get host system: please specify a datacenter```

I've been catching up on things from when I was out. The verbal description I got of this problem from @jak-atx (as a part of a meeting on another topic) suggested that the difficulty here might be related to inability to use a session from a different IP address. It's not clear to be what we know know versus what we're speculating to be true.

This error message seems to be conveying some other cause, but perhaps this is just a bad error message. Has someone investigated where it's coming from?

Have we done some basic testing, such as moving a session key between the UI and govc (without the API being involved) to see if that works? Or using a single UI or govc session from two different IP addresses? It seems like there are a lot of moving parts in the test case above.

cc @hickeng @lcastellano

**Comment by dougm** [comment]
I spoke with @lcastellano about this earlier today and suggested the idea below, where we'd `AcquireCloneTicket` via the UI and use `CloneSession` with that ticket to authenticate:

```go
package main

import (
    "context"
    "fmt"
    "log"

    "github.com/vmware/govmomi/examples"
    "github.com/vmware/govmomi/vim25/methods"
    "github.com/vmware/govmomi/vim25/types"
)

func main() {
    ctx := context.Background()

    c, err := examples.NewClient(ctx) // username/password login happens here
    if err != nil {
        log.Fatal(err)
    }

    var ticket, session string

    {
        // See: http://pubs.vmware.com/vsphere-6-5/topic/com.vmware.wssdk.apiref.doc/vim.SessionManager.html
        // The UI plugin should call this method:
        req := types.AcquireCloneTicket{
            This: *c.ServiceContent.SessionManager,
        }

        res, err := methods.AcquireCloneTicket(ctx, c, &req)
        if err != nil {
            log.Fatalf("AcquireCloneTicket: %s", err)
        }

        ticket = res.Returnval
    }

    // Pass ticket to the service, which would then call:
    {
        req := types.CloneSession{
            This:        *c.ServiceContent.SessionManager,
            CloneTicket: ticket,
        }

        res, err := methods.CloneSession(ctx, c, &req)
        if err != nil {
            log.Fatalf("CloneSession: %s", err)
        }

        session = res.Returnval.Key
        fmt.Printf("session=%s\n", session)
    }
}
```

**Comment by jzt** [comment]
@dougm @lcastellano Thanks guys, this works perfectly. I tagged you both on https://github.com/vmware/vic/pull/6587 so if you have time to give it a quick once-over, I'd be grateful (although it's not much different than this at all).

@dougm I will pull those govmomi vendor changes out once I know this all works (so far, so good) and move it to a govmomi PR.

Thanks again, gentlemen!
